Lint rule for using Ahem as a web font.

LukeZielinski opened the pull request:
Add a lint rule to check for tests that use the Ahem font but do not
include the /fonts/ahem.css stylesheet (which loads this font as a web
font). This prevents tests from assuming that Ahem is installed as a
system font. Also includes some unit tests.

There are a few tests that are incorrectly flagged as violating this
rule, these were added to lint.whitelist.

A bunch of other tests were fixed to load Ahem as a web font.

foolip commented:
Unfortunately this is going to be a bit more complicated as https://github.com/web-platform-tests/wpt/blob/master/css/vendor-imports/mozilla/mozilla-central-reftests/README says "THIS DIRECTORY IS NOT THE MASTER COPY!  DO NOT EDIT TESTS HERE!"

@dbaron @Ms2ger how should we go about making the Ahem changes to those tests?

foolip commented on lint.whitelist:
```diff
@@ -827,3 +827,14 @@ MISSING DEPENDENCY: web-nfc/resources/nfc-helpers.js
+
+# Tests that are false positives for using Ahem as a system font
+AHEM SYSTEM FONT: acid/acid3/test.html
+AHEM SYSTEM FONT: preload/single-download-preload.html
```
This test doesn't even use Ahem, perhaps s/ahem/Foo/ or something to avoid the lint exception?

LukeZielinski replied:
Good point, done.

foolip reviewed (approved):
LGTM % imported tests we can't touch. Can you back out those changes and a lint exception for them while you figure out how to get them into Gecko and back here?

Ms2ger commented:
> Unfortunately this is going to be a bit more complicated as https://github.com/web-platform-tests/wpt/blob/master/css/vendor-imports/mozilla/mozilla-central-reftests/README says "THIS DIRECTORY IS NOT THE MASTER COPY! DO NOT EDIT TESTS HERE!"
> 
> how should we go about making the Ahem changes to those tests?

Update them in mozilla-central, and @dbaron will manually push the changes upstream at some point.

LukeZielinski committed:
Add a lint rule to check for tests that use the Ahem font but do not
include the /fonts/ahem.css stylesheet (which loads this font as a web
font). This prevents tests from assuming that Ahem is installed as a
system font. Also includes some unit tests.

There are a few tests that are incorrectly flagged as violating this
rule, these were added to lint.whitelist.

A bunch of other tests were fixed to load Ahem as a web font.

LukeZielinski reviewed (commented):
> > Unfortunately this is going to be a bit more complicated as https://github.com/web-platform-tests/wpt/blob/master/css/vendor-imports/mozilla/mozilla-central-reftests/README says "THIS DIRECTORY IS NOT THE MASTER COPY! DO NOT EDIT TESTS HERE!"
> > how should we go about making the Ahem changes to those tests?
> 
> Update them in mozilla-central, and @dbaron will manually push the changes upstream at some point.

Thanks, I reverted these. All of these tests did already load Ahem as a web font, just from their own local copies. I think they're fine just being left as is - they do the right thing, and I don't actually see /fonts/ in mozilla-central, so not sure updating them would work.

LukeZielinski commented:
Oh, also, I'm keeping track of the failures introduced by this PR here: crbug.com/1001210
These failures aren't trivially fixed so they will be investigated later.

foolip commented on lint.whitelist:
```diff
@@ -827,3 +827,137 @@ MISSING DEPENDENCY: web-nfc/resources/nfc-helpers.js
+# The following are imported from mozilla-central and can't be modified in WPT.
+# They do load Ahem as a web font, but they use their own copy which trips the
+# lint rule. Basically false positives.
+AHEM SYSTEM FONT: css/vendor-imports/mozilla/mozilla-central-reftests/text3/segment-break-transformation-rules-003.html
```
Since we don't control these tests and the lint isn't run where they are maintained, listing individual tests isn't too useful, and will probably just mean that this list has to be updated on each sync, without driving down the number of exceptions. I recommend a blanket css/vendor-imports/mozilla/mozilla-central-reftests/* here instead.

jgraham replied:
I disagree with this. We should lint those tests like normal since they should be moved out of that directory and we don't want to pile up problems that have to be addressed during that move.

foolip replied:
That's fine with me, it's Mozilla that maintains this. I assumed the wpt lint wouldn't run when modifying these at the upstream location so that it'd just be a burden on @dbaron when syncing, but if that's fine, I'll send a revert of my simplification.

foolip replied:
https://github.com/web-platform-tests/wpt/pull/18929

foolip reviewed (approved):
Suggest simplification to lint.whitelist, but I'll check the CI checks here and if they're good I'll admin merge and do that change as a follow up.

foolip commented:
@LukeZielinski can you also file a Gecko bug for the regressions in https://wpt.fyi/results/?diff&filter=ADC&run_id=306760008&run_id=299020008?

foolip commented:
Sent https://github.com/web-platform-tests/wpt/pull/18901.

gsnedders commented:
@LukeZielinski can you please update the docs as well? we should have the lint rule documented pending #17189 landing, and we shouldn't just say that you can rely on Ahem being installed in various places.

LukeZielinski commented:
@gsnedders Yes, the docs changes were in a pending PR, but I carved them off into #18972 to expedite.
